[storage] Add an SQL storage driver

elafarge opened the pull request:
**What this PR does / why we need it**:

This commits adds the possibility to back Tiller (or the future
Tiller-less Helm CLI) with any SQL database (only postgres has been
tested so far) to store release information.

The main motivation for this commit was to use a storage backend that
would allow releases larger that 1MB in size (ConfigMap or Secret
drivers don't, because of limits on value size in the underlying etcd
key-value store).

Signed-off-by: Étienne Lafarge <etienne.lafarge@gmail.com>

Co-authored-by: Elliot Maincourt <e.maincourt@gmail.com> (@emaincourt)
Co-authored-by: Paul Borensztein <hi@0x01.fr> (@commit-master)

**Special notes for the reviewer(s)**:

* It goes without saying that we'll be glad to maintain and update this part of the codebase - in particular - to comply with the new Tiller-less Helm 3 ;-) 
* Also, we'll be glad to write an `any backend -> postgres` migration script if this backend gets promoted to GA at some point.
* Only `postgresql` has been tested so far. However, we've tried to make our code SQL dialect-agnostic (using `sqlx` and avoiding `postgres` specific SQL statements). We'll be happy to test other databases if required by the community.
* We're currently testing this feature on our pre-production environment (to provide our developers with a "staging-on-demand" feature) and will switch our production infrastructure on it if this PR gets merged

**If applicable**:
- [x] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility --> **not applicable for now, we'll be glad to write a migration script and add it to the helm CLI if needed**

closes #1413
closes #1996 
closes #4215

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/master/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

elafarge committed:
[storage] Add an SQL storage driver

This commits adds the possibility to back Tiller (or the future
Tiller-less Helm CLI) with any SQL database (only postgres has been
tested so far) to store release information.

The main motivation for this commit was to use a storage backend that
would allow releases larger that 1MB in size (ConfigMap or Secret
drivers don't, because of limits on value size in the underlying etcd
key-value store).

Signed-off-by: Étienne Lafarge <etienne.lafarge@gmail.com>

Co-authored-by: Elliot Maincourt <e.maincourt@gmail.com> (@emaincourt)
Co-authored-by: Paul Borensztein <hi@0x01.fr> (@commit-master)

bacongobbler commented on pkg/storage/driver/sql.go:
```diff
@@ -0,0 +1,322 @@
+}
+
+// Release describes a Helm release
+type Release struct {
```
Is there some way we can use the Release object similar to the other storage drivers? I understand the need for the annotations for the sql driver to work, but this means we now have [two different ways of representing Helm releases](https://github.com/helm/helm/blob/33589472cf02d8d10e96e06085d95791c0611a75/pkg/proto/hapi/release/release.pb.go#L19) in the backend, which means that any time we introduce changes to the Release object, we'll also have to update this struct as well.

technosophos replied:
I'm sorta inclined to go this way, since moving the other way would involve adding writing new packages that plug into the protobuf tools, which are already the most fragile piece of Helm. So I would be okay with this... and hopefully Helm 3 will make this much easier.

bacongobbler replied:
Sounds good to me! Thanks for the clarification. :)

bacongobbler commented on cmd/tiller/tiller.go:
```diff
@@ -141,6 +147,18 @@ func start() {
 		env.Releases.Log = newLogger("storage").Printf
+	case storageSQL:
+		sqlDriver, err := driver.NewSQL(
+			*sqlDialect,
```
It would be nice if we validated that the dialect chosen was valid or not. For example, I could enter `postgresql` or `postgres`. Without validation we cannot inform the user which dialects tiller supports or not. As far as I can tell we only accept postgresql as a valid dialect.

elafarge replied:
Totally right, we added some extra validation on the storage driver in the `NewSQL()` method. 

As you mentioned above, since we're importing `pq` only, the current implementation is limited to postgres. That totally makes sense.

bacongobbler commented on docs/sql-storage.md:
```diff
@@ -0,0 +1,89 @@
+You may be willing to store release information in an SQL database - in
+particular, if your releases weigh more than 1MB and therefore [can't be stored in ConfigMaps or Secrets](https://github.com/helm/helm/issues/1413).
+
+We recommend using [PostgreSQL](https://www.postgresql.org/).
```
After looking at the code, it looks like only postgres support is baked in, as it only imports the pq driver, so I'd change this to

> The SQL backend only works with PostgreSQL at this time.

elafarge replied:
:ok_hand:

bacongobbler commented on docs/install.md:
````diff
@@ -369,6 +372,31 @@ Currently, if you want to switch from the default backend to the secrets
+```shell
+helm init \
+  --override \
+    'spec.template.spec.containers[0].args'='{--storage=sql,--sql-dialect=postgres,--sql-connection-string=postgresql://tiller-postgres:5432/helm?user=helm&password=changemeforgodssake&sslmode=disable}'
````
Also would you mind removing the `sslmode` string? In production, one should not be running a postgres database without sslmode configured (or at least the documentation shouldn't be explicitly disabling it).

elafarge replied:
Sure ! Both changes have been made.

elafarge committed:
[pr-review] Lighten docs & validate SQL dialect

Signed-off-by: Étienne Lafarge <etienne.lafarge@gmail.com>

elafarge commented on docs/install.md:
```diff
@@ -369,6 +372,31 @@ Currently, if you want to switch from the default backend to the secrets
+weighs more than 1MB (in which case, it can't be stored in ConfigMaps/Secrets
+because of internal limits in Kubernetes' underlying etcd key-value store).
+
+To enable the SQL backend, you'll need to [deploy an SQL
```
You're totally right, an old habit of mine of pronouncing SQL as "es-queue-ale" instead of "sequel"... 

That's fixed, thanks :)

elafarge commented on docs/sql-storage.md:
```diff
@@ -0,0 +1,89 @@
+# Store release information in an SQL database
```
:ok_hand: . This file was actually deleted and its content left as an exercise for the reader

elafarge commented on docs/sql-storage.md:
```diff
@@ -0,0 +1,89 @@
+# Store release information in an SQL database
+
+You may be willing to store release information in an SQL database - in
```
Thanks for the correction :)

elafarge commented on docs/sql-storage.md:
```diff
@@ -0,0 +1,89 @@
+          - name: POSTGRES_USER
+            value: helm
+          - name: POSTGRES_PASSWORD
+            value: changemeforgodssake
```
:ok_hand:

elafarge commented on docs/sql-storage.md:
```diff
@@ -0,0 +1,89 @@
+
+We recommend using [PostgreSQL](https://www.postgresql.org/).
+
+This document describes how to deploy `postgres` atop Kubernetes. This being
```
That's totally fine to me, we just got rid of this document.

elafarge commented:
Thanks a lot for the quick review @bacongobbler .

All the requested changes have been implemented so far but one.
We'll think about it over the weekend and come back to you on that matter early next week.

Have a great weekend !

technosophos commented on pkg/storage/driver/sql.go:
```diff
@@ -0,0 +1,330 @@
+				Id: "init",
+				Up: []string{
+					`
+            CREATE TABLE releases (
```
Nit: The whitespace here seems to be a mix of spaces and tabs for line prefixing.

technosophos commented on pkg/storage/driver/sql.go:
```diff
@@ -0,0 +1,330 @@
+func (s *SQL) Query(labels map[string]string) ([]*rspb.Release, error) {
+	var sqlFilterKeys []string
+	sqlFilter := map[string]interface{}{}
+	for key, val := range labels {
```
You might need to sanitize the value of `key` to avoid a Bobby Tables attack. https://www.xkcd.com/327/

elafarge replied:
We actually check that `key` is in this (sane) map before executing the generated SQL request: https://github.com/helm/helm/pull/5371/files#diff-9614ccdf050917eda8876d24c7cb63ceR37 at line 197. Would that be fine to you ?

bacongobbler replied:
yeah, that looks fine to me!

technosophos reviewed (changes requested):
I am really excited to see this! I have long wanted to write a driver like this, but have just never had the time/need to do it. I left a security-related comment that I think needs a change (or maybe just an explanation that the SQL driver does something I don't know about). Otherwise, I am good.

Also, thanks for setting a high precedent on how a SQL driver should be tested!

emaincourt committed:
Clarify our SQL Release binding struct naming and purpose

Signed-off-by: Elliot Maincourt <e.maincourt@gmail.com>

elafarge commented:
Thanks a lot for the code review @technosophos . We took all your suggestions into account but the last one. We indeed do not sanitize the SQL field name *but we do check that it belongs to a sane set of possible fields*: https://github.com/helm/helm/pull/5371/files#diff-9614ccdf050917eda8876d24c7cb63ceR37 . 

Regarding the precedent on testing SQL drivers, `go-sqlmock` is pretty useful to write unit tests to test the logic around SQL queries indeed :) 
However, in that case, integration tests are also pretty useful as they check that the SQL backend executes the queries as expected (executes them at all). When we first ran our contribution against a real `postgres` database, we figured out a few typos that haven't been detected by `go-sqlmock` but were obvious when running the queries against a real SQL backend :) 

Is there any plans to add such integration tests to Helm (or do they exist already and we simply didn't find them) ? If so, we'll be glad to add some :)

emaincourt committed:
Fix formatting issue

Signed-off-by: Elliot Maincourt <e.maincourt@gmail.com>

callicles commented:
Thanks for that work! 
I recently run into that was well and I am waiting for that to be able to move to production with Helm

elafarge commented:
@bacongobbler @technosophos All the requested changes have been taken into account. Let us know if there's anything else you'd like us to change / add / delete :) 

Thanks again for your insights and suggestions :)

bacongobbler reviewed (approved):
looks great!

elafarge commented:
@technosophos @bacongobbler we just rebased our branch against `master` and fixed some conflicts in the `glide.lock` file. Let us know if there's anything else you'd like us to change :)

callicles commented:
@technosophos Would be great if that could make it to the next release!

elafarge commented:
(just rebased against the `master` branch, fixed a conflict in `tiller.go` and force-pushed).

tg339 commented:
@technosophos any update on this one :) We're excited to see this come through to help us with getting over a hurdle that's preventing us from using helm in production

bacongobbler commented:
He's currently on vacation for the next week. I'd try asking in #helm-dev on Slack and ask if another maintainer would be willing to review this.

elafarge commented:
(FYI, I just rebased the branch against `master`, fixed conflicts, and force-pushed)

elafarge commented:
Hi @technosophos @bacongobbler ,

Could you take a look at this PR when you get the chance. All required updates have been made. We'll be glad to change/add/remove/discuss anything you feel is relevant in a timely manner so that this can make it to 2.14.0 ;-) 

Have a great day !

technosophos reviewed (approved):
LGTM. I'll leave it to @bacongobbler to merge when it fits the release schedule.

bacongobbler commented:
I completely forgot that we made the call to bring this into the 2.14 release. I'll cut rc.2 with this PR. Thanks @elafarge!

bacongobbler commented:
cherry-picked into 2.14.

bacongobbler commented:
Just a heads up, but the SQL driver tests failed after merging this into master: https://circleci.com/gh/helm/helm/8204

@elafarge would you mind looking into this to fix up the breaking tests?

bacongobbler commented:
Looks like it's due to the underlying gRPC layer being populated. See attached.

![image](https://user-images.githubusercontent.com/1360539/57487527-b37ba900-7265-11e9-8b2f-b7064296bb50.png)

bacongobbler commented:
I've pulled it out of the 2.14 release branch for the time being so we can continue to test the release candidate. @elafarge, when you have time to look into this, please reach out to me and we can see what we can do about getting this in for 2.14. Thanks!

bacongobbler commented:
Found the fix; see #5706.

elafarge commented:
Hi @bacongobbler . Sorry, for the late reply, I was off for the week. Glad to hear you found the fix.

Thanks a lot for the merge !! :tada:

vicaya commented:
Does the sql storage driver work with helm v3?

If not, any plans to port this (and other storage drivers) to v3.

hickeyma commented:
This would ned to be ported to v3. @elafarge Is this something you would be interested in doing?

elafarge commented:
@hickeyma sorry, I just stumbled upon your question when cleaning up my Github notifications. I'll surely take a peek at this (maybe with @emaincourt if he's interested :angel: ) this week.

emaincourt commented:
@elafarge I might actually have time to handle it. Will take a look at the v3 codebase this week 🧚‍♀️

hickeyma commented:
Thanks for getting back @elafarge @emaincourt. That would be great if you can port it. Reach out if you need any help. FYI, Helm v3 branch is now `master`. Push a new PR with a reference this PR.

emaincourt commented:
@hickeyma @elafarge PR here: https://github.com/helm/helm/pull/7635
